### PR TITLE
Prevent invalid column names (className and length)

### PR DESCRIPTION
**Pull request by dplewis**
Fixes: https://github.com/parse-community/parse-server/issues/7047, https://github.com/parse-community/parse-server/issues/6195

There could be more of the reserved fields that could mess up the database or produce unexpected results.

### CODE REVIEW

**Comment by mtrezza** [comment]
Maybe also `authData`?

File: spec/CloudCode.spec.js [hunk 1]
```diff
@@ -216,6 +216,21 @@ describe('Cloud Code', () => {
...
+  it('test beforeSave with invalid field', async () => {
+    Parse.Cloud.beforeSave('BeforeSaveChanged', function (req) {
+      req.object.set('length', 0);
+    });
+
+    const obj = new Parse.Object('BeforeSaveChanged');
+    obj.set('foo', 'bar');
+    try {
+      await obj.save();
+      expect(false).toBe(true);
```
Comment (mtrezza) [inline spec/CloudCode.spec.js:228]: Maybe `fail("should not succeed");`? Should then also be changed in the tests below.

File: spec/ParseQuery.spec.js [hunk 2]
```diff
@@ -2860,31 +2860,17 @@ describe('Parse.Query testing', () => {
     });
   });
 
-  it('object with length', function (done) {
+  it('object with length', async () => {
```
Comment (mtrezza) [inline spec/ParseQuery.spec.js:2863]: Do we need an extra test case for each field name, or does it makes sense to consolidate into a single test case with a loop over an array of field names?

**Review by mtrezza — COMMENTED** [review]
Just cosmetics...

File: spec/ParseQuery.spec.js (hunk 2 above)
Comment (dplewis) [reply spec/ParseQuery.spec.js]: Done

**Review by mtrezza — APPROVED** [review]
LGTM 🚀

**Comment by penchef** [comment]
Hey, 

so it turns out preventing the column name `length` *is* a breaking change (at least for some people). Having this breaking change wrapped up in the security relevant 4.5.0 release makes it a huge problem to stay on top of security issues. It would be nice to
1.  allow `length` again for a 4.5.1 release
2. re-introduce this in a subsequent PR + [document|https://parseplatform.org/Parse-SDK-JS/api/master/Parse.Object.html] the prohibit column names

I'd suggest to rename forbidden fields to `_length`, `_className` etc. like parse does with the `_User` class. Adding a `length` field is not that unlikely.

**Comment by pausenabre** [comment]
What is the work around for this breaking change? We are currently using classes in production with the `length` field. 

Do we need to do a migration of the PFObjects that currently use the `length` field? 
at the moment attempting to save objects with `length` prompts a Error 105, and we can not delete the field neither as we get the same error: 

`Error Domain=Parse Code=105 "Invalid field name: length." UserInfo={error=Invalid field name: length., NSLocalizedDescription=Invalid field name: length., code=105}`

I would also suggest to allow the use of `length` again and some documentation on how to migrate existing classes using `length` or any other prohibit names.

**Comment by julianvogels** [comment]
To add to this (@pausenabre is my colleague): 
We are using the a `length` column on one of our Parse classes. Haven't been aware that this is a restricted field. The issues we're seeing after upgrading to Parse Server 4.5.0 are that we can't save any objects with the `length` property set (from iOS). 

As a workaround, we were trying to `unset` the `length` key in a `beforeSave` hook. That works great for any object which is newly created, but not for object _updates_ (tested with REST API `POST` and `PUT`). When updating an object, the `beforeSave` hook is not called at all and Parse Server returns an error before I can `unset` the key. So the first solution would be to call the `beforeSave` hook for updates before returning the error about the `length` key.

We are also using the Parse iOS SDK local datastore, where in our production app, users have already saved objects which contain data for the `length` key. Once a user saves the local object to the server, they would see this error. We were not able to find a way to `unset` the `length` key client-side on iOS. That would be the second solution to the issue.

**Comment by mtrezza** [comment]
I agree that allowing virtually any keyword as field name would be reasonable. The currently "reserved" field names could be allowed by encoding fields differently in the Parse Object.

**Pull request merged**